### PR TITLE
make projectile predicate getter sensitive to weapon stack in ProjectileWeaponItem

### PR DESCRIPTION
**Pull request by lcy0x1**
Currently `ProjectileWeaponItem::getAllSupportedProjectiles` is not sensitive to weapon stack, making it difficult to implement certain weapons that takes ammunition based on stack data. This implementation is aimed to fix that issue.

### CODE REVIEW

File: patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch [hunk 1]
```diff
@@ -1,5 +1,34 @@
...
+         return this.getAllSupportedProjectiles();
+     }
+ 
++    /**
++     * Use ItemStack sensitive version instead
++     */
++    @Deprecated
+     public abstract Predicate<ItemStack> getAllSupportedProjectiles();
+ 
++    public Predicate<ItemStack> getSupportedHeldProjectiles(ItemStack stack) {
```
Comment (HenryLoenwind) [inline patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch:29]: Please add some javadoc describing what this should do and what the stack parameter is.

File: patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch [hunk 2]
```diff
@@ -1,5 +1,34 @@
 --- a/net/minecraft/world/item/ProjectileWeaponItem.java
 +++ b/net/minecraft/world/item/ProjectileWeaponItem.java
+@@ -24,12 +_,28 @@
+         super(p_43009_);
+     }
+ 
++    /**
++     * Use ItemStack sensitive version instead
++     */
++    @Deprecated
```
Comment (HenryLoenwind) [inline patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch:10]: Please also use the javadoc deprecated marker; it shows up as part of the mouse-over warning, while the javadoc itself may be hidden by the warning.

File: patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch (hunk 1 above)
Comment (lcy0x1) [reply patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch:29]: resolved in commit #a091806

File: patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch (hunk 2 above)
Comment (lcy0x1) [reply patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch:10]: resolved in commit #a091806

**Comment by lcy0x1** [comment]
Added more javadoc in commit #a091806

**Review by HenryLoenwind — APPROVED** [review]

File: patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch [hunk 3]
```diff
@@ -1,5 +1,49 @@
...
++     */
++    public Predicate<ItemStack> getSupportedHeldProjectiles(ItemStack stack) {
++        return getAllSupportedProjectiles(stack).or(getSupportedHeldProjectiles());
++    }
++
++    /**
++     * Override this method if the allowed projectile is weapon stack dependent.
++     *
++     * @param stack The ProjectileWeapon stack
++     * @return A predicate that returns true for all supported projectile stack regardless of position
```
Comment (TelepathicGrunt) [inline patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch]: What does position mean here?

File: patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch [hunk 4]
```diff
@@ -1,5 +1,49 @@
...
++    /**
++     * @deprecated Use ItemStack sensitive version {@link ProjectileWeaponItem#getSupportedHeldProjectiles(ItemStack)}
++     */
++    @Deprecated
+     public Predicate<ItemStack> getSupportedHeldProjectiles() {
+         return this.getAllSupportedProjectiles();
+     }
+ 
++    /**
++     * If the allowed projectile is weapon stack dependent, return {@code ItemStack.EMPTY}
```
Comment (TelepathicGrunt) [inline patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch]: Is encouraging people to return empty a good idea here? Wouldn't that break all mods currently using this deprecated method that calls an updated mod's method? Might be better to not say to return empty and let the mod determine if they do want to return empty or some default ammo for legacy callers

File: patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch (hunk 3 above)
Comment (lcy0x1) [reply patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch]: Position here means where the projectile stack is: in hand, in inventory, etc. This is opposed to `getSupportedHeldProjectiles` where only hand is checked.

File: patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch (hunk 4 above)
Comment (lcy0x1) [reply patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch]: Sure. Let me change that

File: patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch (hunk 3 above)
Comment (TelepathicGrunt) [reply patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch]: There must be a better way to describe that as it is a bit ambiguous rn. Could just say `supported projectile stack regardless of what slot the projectiles are in`

Comment (lcy0x1) [reply patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch]: Yeah I figured that's unclear. Is it better if I just say `supported projectile stack`? I think we could expect people who are overriding `ProjectileWeaponStack` to read javadoc over `getSupportedHeldProjectile` and understand that the other method supports only projectile stacks on hand

Comment (lcy0x1) [reply patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch]: JavaDoc updated

**Review by marchermans — APPROVED** [review]

**Pull request merged**